### PR TITLE
[Diffusion] LTX-2 Support PR1

### PR DESCRIPTION
**Pull request by gmixiaojin**
# Motivation
Support **LTX-2 Video & Audio Joint** model.

# Contribution
- Modify entrypoints to support audio
- Modify RoPE to support LTX-2's RoPE mechanism
- Modify the component loader to load LTX-2's model parameter
- Modify pipelines_core stages to support audio

# Review Process
1. Ping **Merge Oncalls** to start the PR flow. See **[PR Merge Process]**.
2. Get approvals from **CODEOWNERS** and other reviewers.
3. Trigger CI tests with comments or contact authorized users to do so.
   - `\tag-run-ci-label`, `\rerun-failed-ci`, `\tag-and-rerun-ci`
4. After green CI and required approvals, ask **Merge Oncalls** to merge.

### CODE REVIEW

File: python/sglang/multimodal_gen/runtime/entrypoints/openai/utils.py [hunk 1]
```diff
@@ -216,17 +217,27 @@ async def process_generation_batch(
...
             )
         save_file_path_list = []
+        audio_sample_rate = getattr(result, "audio_sample_rate", None)
         if batch.data_type == DataType.VIDEO:
             for idx, output in enumerate(result.output):
                 save_file_path = str(
                     os.path.join(batch.output_path, batch.output_file_name)
                 )
+                sample = result.output[idx]
+                audio = getattr(result, "audio", None)
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/entrypoints/openai/utils.py]: please avoid using `getattr` as much as possible, in all places

File: python/sglang/multimodal_gen/runtime/entrypoints/diffusion_generator.py [hunk 2]
```diff
@@ -231,8 +235,27 @@ def generate(
                             request_idx + 1,
                         )
                         continue
+                    audio_sample_rate = getattr(output_batch, "audio_sample_rate", None)
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/entrypoints/diffusion_generator.py]: seems duplicate with the previous one, could you unify them as a follow-up?

File: python/sglang/multimodal_gen/runtime/entrypoints/utils.py [hunk 3]
```diff
@@ -86,6 +113,138 @@ def post_process_sample(
...
+                                    inferred_sr = int(
+                                        round(float(audio_len) / duration_s)
+                                    )
+                                    if 8000 <= inferred_sr <= 192000:
+                                        selected_sr = inferred_sr
+                            except Exception:
+                                pass
+
+                        try:
+                            import shutil
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/entrypoints/utils.py]: move imports to the top of this file

File: python/sglang/multimodal_gen/runtime/entrypoints/utils.py [hunk 4]
```diff
@@ -86,6 +113,138 @@ def post_process_sample(
                     codec="libx264",
                     quality=quality,
                 )
+
+                if audio is not None:
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/entrypoints/utils.py]: extract the audio save logic to a helper

File: python/sglang/multimodal_gen/runtime/loader/component_loader.py [hunk 5]
```diff
@@ -755,6 +832,58 @@ def load_customized(
         return model
 
 
+class AdapterLoader(ComponentLoader):
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/loader/component_loader.py:847]: this file is too large, consider splitting it into a folder `component_loader` in a follow-up

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/text_connector.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
+import torch
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/text_connector.py]: we could use a better file name

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising_av.py [hunk 7]
```diff
@@ -0,0 +1,729 @@
...
+                device=latents.device,
+                dtype=torch.float32,
+            )
+            denoise_mask[:, :num_img_tokens, :] = 0.0
+            clean_latent = latents.detach().clone()
+            clean_latent[:, :num_img_tokens, :] = batch.image_latent[
+                :, :num_img_tokens, :
+            ].to(device=latents.device, dtype=latents.dtype)
+
+        with torch.autocast(
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising_av.py]: do we have a plan to reduce duplication code here?

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/upsampling.py [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+
+
+class LTX2UpsamplingStage(PipelineStage):
+    def __init__(self, upsampler, video_encoder_stats):
+        super().__init__()
+        self.upsampler = upsampler
+        self.video_encoder_stats = video_encoder_stats
+
+    def forward(self, batch: Req, server_args: ServerArgs) -> Req:
+        if hasattr(batch, "latents") and batch.latents is not None:
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/upsampling.py]: in what case do we not have a `latents` in batch?

File: python/sglang/multimodal_gen/runtime/pipelines_core/schedule_batch.py [hunk 9]
```diff
@@ -82,11 +82,25 @@ class Req:
...
     latents: torch.Tensor | None = None
     # Flux-2
     latent_ids: torch.Tensor | None = None
 
+    # Audio Latents (LTX-2)
+    audio_latents: torch.Tensor | None = None
+    raw_audio_latent_shape: tuple[int, ...] | None = None
+
+    # Audio Parameters
+    frame_rate: float = 24.0
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/pipelines_core/schedule_batch.py]: could use fps here

**Review by mickqian — CHANGES_REQUESTED** [review]

File: python/sglang/multimodal_gen/runtime/entrypoints/utils.py (hunk 4 above)
Comment (gmixiaojin) [reply python/sglang/multimodal_gen/runtime/entrypoints/utils.py]: Refact the code and resolved.

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising_av.py (hunk 7 above)
Comment (gmixiaojin) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising_av.py]: Yes, we suggest that we may refact this one after the PR?

File: python/sglang/multimodal_gen/runtime/loader/component_loader.py (hunk 5 above)
Comment (gmixiaojin) [reply python/sglang/multimodal_gen/runtime/loader/component_loader.py:847]: Sure, I think we will follow-up after the PR

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/text_connector.py (hunk 6 above)
Comment (gmixiaojin) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/text_connector.py]: changed to text_connector

File: python/sglang/multimodal_gen/runtime/pipelines_core/schedule_batch.py (hunk 9 above)
Comment (gmixiaojin) [reply python/sglang/multimodal_gen/runtime/pipelines_core/schedule_batch.py]: fixed.

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/upsampling.py (hunk 8 above)
Comment (gmixiaojin) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/upsampling.py]: file deleted

**Review by mickqian — APPROVED** [review]

File: python/sglang/cli/utils.py [hunk 10]
```diff
@@ -90,7 +90,7 @@ def _maybe_download_model(
 
 
 # Copied and adapted from hf_diffusers_utils.py
-def is_diffusers_model_path(model_path: str) -> True:
+def is_diffusers_model_path(model_path: str) -> bool:
```
Comment (mickqian) [inline python/sglang/cli/utils.py]: please don't change this file in this PR

Comment (gmixiaojin) [reply python/sglang/cli/utils.py]: resolved.
